### PR TITLE
Fix rare synchronization issue in accountUpdates

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary

We have found a rare synchronization issue in accountUpdates. The issue is expressed when querying for a balance beyond the MaxBalLookback boundaries, where is can result in providing an incorrect result.

The expected result is either an error, saying that the information is not available ( i.e. `round 2 before dbRound 5` or such ), or ( in case we haven't flushed the information to disk) we could return the account data.
In the case we return the account data, it need to match the requested round.

Some special handling was provided for shared cache connections, as these have a different locking semantics compared to a standard disk, private cache connections.

## Testing

A unit test was added to provide validation for the fix. Unit test fails prior to the fix, and passes with the fix.

The in-memory test always fail without the fix. The on-disk test randomly fails. To make the on-disk fail more consistent ( without the fix, naturally), one would need to add a sleep in the commitRound, just after the Atomic call.

### CODE REVIEW

File: ledger/acctupdates_test.go [hunk 1]
```diff
@@ -596,3 +597,157 @@ func TestLargeAccountCountCatchpointGeneration(t *testing.T) {
...
 }
+
+// The TestAcctUpdatesUpdatesCorrectness conduct a correctless test for the accounts update in the following way -
+// Each account is initialized with 100 algos.
+// On every round, each account move variable amount of funds to an accumulating account.
+// The deltas for each accounts are picked by using the lookup method.
+// At the end of the test, we verify that each account has the expected amount of algos.
+// In addition, throughout the test, we check ( using lookup ) that the historical balances, *beyond* the
+// lookback are generating either an error, or returning the correct amount.
+func TestAcctUpdatesUpdatesCorrectness(t *testing.T) {
```
Comment (algorandskiy) [inline ledger/acctupdates_test.go:608]: does it constantly reproduce the bug?

File: ledger/acctupdates.go [hunk 2]
```diff
@@ -1008,7 +1008,20 @@ func (au *accountUpdates) commitRound(offset uint64, dbRound basics.Round, lookb
 	var catchpointLabel string
 	beforeUpdatingBalancesTime := time.Now()
 	var trieBalancesHash crypto.Digest
+
+	// the lockTaken variable would help us to ensure that we take the lock before commiting the transaction to the database.
+	// since the function within the "Atomic" migth be re-attempted, we need to perform the check at the begining, and after
+	// we exit from Atomic. Note that there is no concurrently issue within this function around testing the variable value.
```
Comment (algorandskiy) [inline ledger/acctupdates.go:1014]: I guess more explanation needed - atomic function might be successful but re-attempted if commit inside atomic failed.

File: ledger/acctupdates.go [hunk 3]
```diff
@@ -1045,10 +1058,22 @@ func (au *accountUpdates) commitRound(offset uint64, dbRound basics.Round, lookb
...
+			lockTaken = true
+		}
 		return err
 	})
 
 	if err != nil {
+		// if we failed during the commit, undo the lock.
+		if lockTaken {
+			au.accountsMu.Unlock()
+		}
```
Comment (algorandskiy) [inline ledger/acctupdates.go:1075]: can't it be a defer func?

File: ledger/acctupdates.go [hunk 4]
```diff
@@ -1122,7 +1146,8 @@ func (au *accountUpdates) commitRound(offset uint64, dbRound basics.Round, lookb
 	au.accountsMu.Unlock()
```
Comment (algorandskiy) [inline ledger/acctupdates.go:1146]: add a warning if not lockTaken? It should not happen but might serve as kind of protection for future code changes.

Comment (algomaxj) [reply ledger/acctupdates.go:1146]: yeah, I agree -- I think a check just after error checking the `Atomic` call would be a good sanity check.

File: ledger/acctupdates.go [hunk 5]
```diff
@@ -1045,10 +1058,22 @@ func (au *accountUpdates) commitRound(offset uint64, dbRound basics.Round, lookb
...
+			au.accountsMu.Lock()
+			lockTaken = true
+		}
 		return err
 	})
 
 	if err != nil {
+		// if we failed during the commit, undo the lock.
+		if lockTaken {
+			au.accountsMu.Unlock()
```
Comment (algomaxj) [inline ledger/acctupdates.go:1074]: `lockTaken = false`

File: ledger/acctupdates_test.go (hunk 1 above)
Comment (tsachiherman) [reply ledger/acctupdates_test.go:608]: On my laptop, this unit test consistently reproducing the bug. Note that it's using scenarios which aren't exercised in algod - i.e. looking back beyond the loopback window, having a very short loopback window to allow changes to buffer, etc.

File: ledger/acctupdates.go (hunk 4 above)
Comment (tsachiherman) [reply ledger/acctupdates.go:1146]: I can't see any reason not to add it.

File: ledger/acctupdates.go (hunk 3 above)
Comment (tsachiherman) [reply ledger/acctupdates.go:1075]: it depends where you want to defer this function. In general, no - we want to generate the catchpoint without having the lock taken. Adding a defer function inside this if branch is doable - but not very useful : we're exiting from the function few lines after.

File: ledger/acctupdates.go (hunk 2 above)
Comment (tsachiherman) [reply ledger/acctupdates.go]: updated.

File: ledger/acctupdates.go (hunk 5 above)
Comment (tsachiherman) [reply ledger/acctupdates.go:1076]: why ? we're exiting from commitRound, why should we care about local variables anymore ?

Comment (algomaxj) [reply ledger/acctupdates.go:1076]: for the purpose of the check you just added, the variable should stay up to date

Comment (algomaxj) [reply ledger/acctupdates.go:1076]: but I suppose it doesn't actually matter.

File: ledger/acctupdates.go [hunk 6]
```diff
@@ -1045,14 +1060,33 @@ func (au *accountUpdates) commitRound(offset uint64, dbRound basics.Round, lookb
 				return
 			}
 		}
+
```
Comment (algomaxj) [inline ledger/acctupdates.go:1063]: this isn't from this PR, but it looks like the return value from `updateAccountsRound` is getting ignored above if `isCatchpointRound` is true.

Comment (tsachiherman) [reply ledger/acctupdates.go:1063]: That should be fixed.. I'll add this to this PR.

**Comment by algorandskiy** [comment]
From the Travis log:
```
POTENTIAL DEADLOCK:
Previous place where the lock was grabbed
goroutine 1151560 lock 0xc0003f0ba8
acctupdates.go:253 ledger.(*accountUpdates).lookup { au.accountsMu.RLock() } <<<<<
acctupdates_test.go:711 ledger.TestAcctUpdatesUpdatesCorrectness { acct, err := au.lookup(basics.Round(checkRound), moneyAccounts[j], false) }

Have been trying to lock it again for more than 30s
goroutine 1151587 lock 0xc0003f0ba8
acctupdates.go:1069 ledger.(*accountUpdates).commitRound.func2 { au.accountsMu.Lock() } <<<<<
../util/db/dbutil.go:205 db.(*Accessor).Atomic.func1 { err = fn(tx) }
../util/db/dbutil.go:251 db.(*Accessor).Atomic { err = guardedFn(tx) }
acctupdates.go:1019 ledger.(*accountUpdates).commitRound { err := au.dbs.wdb.Atomic(func(tx *sql.Tx) (err error) { }
acctupdates.go:922 ledger.(*accountUpdates).commitSyncer { au.commitRound(committedOffset.offset, committedOffset.dbRound, committedOffset.lookback) }

Here is what goroutine 1151560 doing now
goroutine 1151560 [chan receive, locked to thread]:
github.com/mattn/go-sqlite3.unlock_notify_wait(0x7f4ce0043a58, 0x0)
	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3_opt_unlock_notify.go:90 +0x16c
github.com/mattn/go-sqlite3._cgoexpwrap_210ed2050b13_unlock_notify_wait(0x7f4ce0043a58, 0xc03da3f378)
	_cgo_gotypes.go:1341 +0x39
github.com/mattn/go-sqlite3._Cfunc__sqlite3_step_internal(0x7f4ce0025ee8, 0x7f4c00000000)
	_cgo_gotypes.go:352 +0x6b
github.com/mattn/go-sqlite3.(*SQLiteRows).Next.func1(0xc006caff90, 0x15ab910)
	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1983 +0x9d
github.com/mattn/go-sqlite3.(*SQLiteRows).Next(0xc006caff90, 0xc03dc53180, 0x1, 0x1, 0x0, 0x0)
	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1983 +0x110
database/sql.(*Rows).nextLocked(0xc03da79800, 0x450000)
	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:2680 +0x155
database/sql.(*Rows).Next.func1()
	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:2658 +0x58
database/sql.withLock(0x18d9c60, 0xc03da79830, 0xc03da3f698)
	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:3097 +0x75
database/sql.(*Rows).Next(0xc03da79800, 0x15a8840)
	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:2657 +0x96
database/sql.(*Row).Scan(0xc03dc8f400, 0xc03da3f7b8, 0x1, 0x1, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:3042 +0x193
github.com/algorand/go-algorand/ledger.(*accountsDbQueries).lookup.func1(0x1171c22, 0x118029c)
	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/accountdb.go:408 +0x1be
github.com/algorand/go-algorand/util/db.LoggedRetry(0xc03dc8f360, 0x18fe900, 0xc000190a70, 0x0, 0xc03da3f8c8)
	/home/travis/gopath/src/github.com/algorand/go-algorand/util/db/dbutil.go:158 +0x52
github.com/algorand/go-algorand/util/db.Retry(...)
	/home/travis/gopath/src/github.com/algorand/go-algorand/util/db/dbutil.go:167
github.com/algorand/go-algorand/ledger.(*accountsDbQueries).lookup(0xc000b39f20, 0x6623f10f3b0a3247, 0xcd1d16b8492380f9, 0x4fc9bc9c5b86bc3, 0xedc467be396899c2, 0xc03da40030, 0x467671, 0x122d560, 0x7f4cc8bfdb28, 0xc00000ea20, ...)
	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/accountdb.go:406 +0x182
github.com/algorand/go-algorand/ledger.(*accountUpdates).lookup(0xc0003f0800, 0x19, 0x6623f10f3b0a3247, 0xcd1d16b8492380f9, 0x4fc9bc9c5b86bc3, 0xedc467be396899c2, 0x1561400, 0x0, 0x0, 0x0, ...)
	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/acctupdates.go:294 +0x3ce
github.com/algorand/go-algorand/ledger.TestAcctUpdatesUpdatesCorrectness(0xc000118900)
	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/acctupdates_test.go:711 +0x184f
testing.tRunner(0xc000118900, 0x15a9478)
FAIL	github.com/algorand/go-algorand/ledger	354.293s
```

**Comment by tsachiherman** [comment]
> From the Travis log:
> 
> ```
> POTENTIAL DEADLOCK:
> Previous place where the lock was grabbed
> goroutine 1151560 lock 0xc0003f0ba8
> acctupdates.go:253 ledger.(*accountUpdates).lookup { au.accountsMu.RLock() } <<<<<
> acctupdates_test.go:711 ledger.TestAcctUpdatesUpdatesCorrectness { acct, err := au.lookup(basics.Round(checkRound), moneyAccounts[j], false) }
> 
> Have been trying to lock it again for more than 30s
> goroutine 1151587 lock 0xc0003f0ba8
> acctupdates.go:1069 ledger.(*accountUpdates).commitRound.func2 { au.accountsMu.Lock() } <<<<<
> ../util/db/dbutil.go:205 db.(*Accessor).Atomic.func1 { err = fn(tx) }
> ../util/db/dbutil.go:251 db.(*Accessor).Atomic { err = guardedFn(tx) }
> acctupdates.go:1019 ledger.(*accountUpdates).commitRound { err := au.dbs.wdb.Atomic(func(tx *sql.Tx) (err error) { }
> acctupdates.go:922 ledger.(*accountUpdates).commitSyncer { au.commitRound(committedOffset.offset, committedOffset.dbRound, committedOffset.lookback) }
> 
> Here is what goroutine 1151560 doing now
> goroutine 1151560 [chan receive, locked to thread]:
> github.com/mattn/go-sqlite3.unlock_notify_wait(0x7f4ce0043a58, 0x0)
> 	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3_opt_unlock_notify.go:90 +0x16c
> github.com/mattn/go-sqlite3._cgoexpwrap_210ed2050b13_unlock_notify_wait(0x7f4ce0043a58, 0xc03da3f378)
> 	_cgo_gotypes.go:1341 +0x39
> github.com/mattn/go-sqlite3._Cfunc__sqlite3_step_internal(0x7f4ce0025ee8, 0x7f4c00000000)
> 	_cgo_gotypes.go:352 +0x6b
> github.com/mattn/go-sqlite3.(*SQLiteRows).Next.func1(0xc006caff90, 0x15ab910)
> 	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1983 +0x9d
> github.com/mattn/go-sqlite3.(*SQLiteRows).Next(0xc006caff90, 0xc03dc53180, 0x1, 0x1, 0x0, 0x0)
> 	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1983 +0x110
> database/sql.(*Rows).nextLocked(0xc03da79800, 0x450000)
> 	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:2680 +0x155
> database/sql.(*Rows).Next.func1()
> 	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:2658 +0x58
> database/sql.withLock(0x18d9c60, 0xc03da79830, 0xc03da3f698)
> 	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:3097 +0x75
> database/sql.(*Rows).Next(0xc03da79800, 0x15a8840)
> 	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:2657 +0x96
> database/sql.(*Row).Scan(0xc03dc8f400, 0xc03da3f7b8, 0x1, 0x1, 0x0, 0x0)
> 	/home/travis/.gimme/versions/go1.12.17.linux.amd64/src/database/sql/sql.go:3042 +0x193
> github.com/algorand/go-algorand/ledger.(*accountsDbQueries).lookup.func1(0x1171c22, 0x118029c)
> 	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/accountdb.go:408 +0x1be
> github.com/algorand/go-algorand/util/db.LoggedRetry(0xc03dc8f360, 0x18fe900, 0xc000190a70, 0x0, 0xc03da3f8c8)
> 	/home/travis/gopath/src/github.com/algorand/go-algorand/util/db/dbutil.go:158 +0x52
> github.com/algorand/go-algorand/util/db.Retry(...)
> 	/home/travis/gopath/src/github.com/algorand/go-algorand/util/db/dbutil.go:167
> github.com/algorand/go-algorand/ledger.(*accountsDbQueries).lookup(0xc000b39f20, 0x6623f10f3b0a3247, 0xcd1d16b8492380f9, 0x4fc9bc9c5b86bc3, 0xedc467be396899c2, 0xc03da40030, 0x467671, 0x122d560, 0x7f4cc8bfdb28, 0xc00000ea20, ...)
> 	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/accountdb.go:406 +0x182
> github.com/algorand/go-algorand/ledger.(*accountUpdates).lookup(0xc0003f0800, 0x19, 0x6623f10f3b0a3247, 0xcd1d16b8492380f9, 0x4fc9bc9c5b86bc3, 0xedc467be396899c2, 0x1561400, 0x0, 0x0, 0x0, ...)
> 	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/acctupdates.go:294 +0x3ce
> github.com/algorand/go-algorand/ledger.TestAcctUpdatesUpdatesCorrectness(0xc000118900)
> 	/home/travis/gopath/src/github.com/algorand/go-algorand/ledger/acctupdates_test.go:711 +0x184f
> testing.tRunner(0xc000118900, 0x15a9478)
> FAIL	github.com/algorand/go-algorand/ledger	354.293s
> ```

Not sure why it's happening. I wasn't able to reproduce it locally.

**Comment by tsachiherman** [comment]
> Looks good to me.
> One small improvement idea: move all lock decision making business to `CondWriteLocker` class and use `condLocker.Lock()` and `condLock.Unlock()`

Sounds like a worthy goal. I'll look into that.

File: util/db/dbutil.go [hunk 7]
```diff
@@ -170,7 +172,7 @@ func Retry(fn func() error) (err error) {
 // getDecoratedLogger retruns a decorated logger that includes the readonly true/false, caller and extra fields.
 func (db *Accessor) getDecoratedLogger(fn idemFn, extras ...interface{}) logging.Logger {
 	log := db.logger().With("readonly", db.readOnly)
-	_, file, line, ok := runtime.Caller(2)
+	_, file, line, ok := runtime.Caller(3)
```
Comment (tsachiherman) [inline util/db/dbutil.go:175]: We increase this to 3 since the "depth" is now longer - we have the "Atomic" function calling the "atomic" function.

**Pull request merged**